File: Baekjoon/Problem_14000~14999/Problem_14891/Problem_14891.cpp
#include <cstdio>
#include <deque>
#include <utility>

enum { N, S };
enum { COUNTERCLOCKWISE=-1, CLOCKWISE=1 };

using namespace std;
deque<int> wheels[4] = { {}, {}, {}, {} };

void move(const int start, const int dir) {
    deque<pair<int, int> > q;
    bool visited[4] = { false };
    q.push_back({ start, dir });

    while (!q.empty()) {
        int cur = q.front().first, dir = q.front().second;
        q.pop_front();
        int left = cur - 1, right = cur + 1;

        if (0 <= left && !visited[left] && wheels[left][2] != wheels[cur][6]) {
            q.push_back({ left, -dir });
        }
        if (right < 4 && !visited[right] && wheels[cur][2] != wheels[right][6]) {
            q.push_back({ right, -dir });
        }

        if (dir == CLOCKWISE) {
            wheels[cur].push_front(wheels[cur].back());
            wheels[cur].pop_back();
        } else {
            wheels[cur].push_back(wheels[cur].front());
            wheels[cur].pop_front();
        }
        visited[cur] = true;
    }
}

int main() {
    for (int i = 0; i < 4; i++) {
        char str[9]; scanf("%s", str);

        for (int j = 0; j < 8; j++) {
            wheels[i].push_back(str[j] == '1' ? 1 : 0);
        }
    }

    int K; scanf("%d", &K);
    for (int t=0; t < K; t++) {
        int i, dir; scanf("%d %d", &i, &dir);

        move(i-1, dir);
    }

    int score = 0;
    score += wheels[0][0] ? 1 : 0;
    score += wheels[1][0] ? 2 : 0;
    score += wheels[2][0] ? 4 : 0;
    score += wheels[3][0] ? 8 : 0;

    printf("%d\n", score);
    
    return 0;
}

File: Baekjoon/Problem_14000~14999/Problem_14891/Problem_14891.py
import sys
from collections import deque

CLOCKWISE, COUNTERCLOCKWISE = 1, -1
N, S = 0, 1

input = lambda: sys.stdin.readline().rstrip()

def move(start, dir):
    global wheels
    queue = deque()
    visited = [ False ] * 4
    queue.append((start, dir))
    
    while queue:
        cur, dir = queue.popleft()
        left, right = cur-1, cur+1
        
        if 0 <= left and not visited[left] and wheels[left][2] != wheels[cur][6]:
            queue.append((left, -dir))
        if right < 4 and not visited[right] and wheels[cur][2] != wheels[right][6]:
            queue.append((right, -dir))
        
        if dir == CLOCKWISE:
            wheels[cur].appendleft(wheels[cur].pop())
        else:
            wheels[cur].append((wheels[cur].popleft()))
        visited[cur] = True

if __name__ == "__main__":
    wheels = [ deque([int(e) for e in input()]) for _ in range(4) ]
    
    K = int(input())
    for _ in range(K):
        i, dir = map(int, input().split())
        move(i-1, dir)
    
    score = 0
    score += 1 if wheels[0][0] else 0
    score += 2 if wheels[1][0] else 0
    score += 4 if wheels[2][0] else 0
    score += 8 if wheels[3][0] else 0
    print(score)
        